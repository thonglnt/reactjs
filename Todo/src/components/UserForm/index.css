.user-form {
  display: grid;
  justify-content: center;
  margin-top: 100px;
}

.user-name,
.avatar-url {
  padding-bottom: 20px;
}

.cancel {
  margin-right: 10px;
  text-decoration: none;
}

.button-wrapper {
  display: flex;
  justify-content: center;
}
