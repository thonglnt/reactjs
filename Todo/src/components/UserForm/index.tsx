// Lib
import { FormEvent } from 'react';
import { Link } from 'react-router-dom';

// Components
import Input from '@components/Input';
import { Spinner } from '@components/Spinner';

// Constants
import { PATH } from '@constants/paths';

export type UserFormProps = {
  onRegisterUser: (event: FormEvent<HTMLFormElement>) => void;
  formRef: React.LegacyRef<HTMLFormElement> | null;
  isSubmitting: boolean;
};

const UserForm = ({ onRegisterUser, formRef, isSubmitting = false }: UserFormProps) => {
  return (
    <form
      method='post'
      ref={formRef}
      data-testid='user-form'
      className='user-form'
      onSubmit={onRegisterUser}
    >
      <h2 className='title'>Create new user</h2>
      <div className='user-name'>
        <Input id='name' name='name' label='Name' placeholder='Enter name' minLength={6} />
      </div>
      <div className='avatar-url'>
        <Input
          id='avatar'
          type='url'
          name='avatar'
          label='Avatar'
          placeholder='Enter link avatar'
        />
      </div>
      <div className='button-wrapper'>
        <Link className='btn-secondary cancel' to={PATH.HOME}>
          Cancel
        </Link>
        <button type='submit' className='btn-success'>
          {isSubmitting ? <Spinner className='progress' /> : 'Submit'}
        </button>
      </div>
    </form>
  );
};

export default UserForm;
