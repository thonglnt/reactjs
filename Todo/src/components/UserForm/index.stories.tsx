// Lib
import { Story } from '@storybook/react';
import { action } from '@storybook/addon-actions';

// Components
import UserForm, { UserFormProps } from '@components/UserForm';
import { BrowserRouter } from 'react-router-dom';

export default {
  title: 'UserForm',
  component: UserForm,
};

export const UserFormSample: Story<UserFormProps> = () => (
  <BrowserRouter>
    <UserForm onRegisterUser={action('onRegisterUser')} formRef={null} isSubmitting={false} />
  </BrowserRouter>
);
