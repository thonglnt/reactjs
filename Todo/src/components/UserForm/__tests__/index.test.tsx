// Lib
import { render, screen, fireEvent } from '@testing-library/react';
import { BrowserRouter } from 'react-router-dom';

// Components
import UserForm, { UserFormProps } from '@components/UserForm';

describe('Test UserForm component', () => {
  const initUserForm: UserFormProps = {
    onRegisterUser: jest.fn(),
    formRef: null,
    isSubmitting: false
  };

  const customUserFormProps: UserFormProps = {
    onRegisterUser: jest.fn(),
    formRef: null,
    isSubmitting: true
  };

  test('It should render correctly with initial Props', () => {
    const { container } = render(
      <BrowserRouter>
        <UserForm {...initUserForm} />
      </BrowserRouter>,
    );

    expect(container).toMatchSnapshot();
  });

  test('It should render correctly with custom User Form Props', () => {
    const { container } = render(
      <BrowserRouter>
        <UserForm {...customUserFormProps} />
      </BrowserRouter>,
    );

    expect(container).toMatchSnapshot();
  });

  test('It should call on register user', () => {
    render(
      <BrowserRouter>
        <UserForm {...initUserForm} />
      </BrowserRouter>,
    );
    fireEvent.submit(screen.getByTestId('user-form'));

    expect(initUserForm.onRegisterUser).toHaveBeenCalledTimes(1);
  });
});
