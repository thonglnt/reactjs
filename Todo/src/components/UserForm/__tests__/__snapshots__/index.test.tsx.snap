// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test UserForm component It should render correctly with custom User Form Props 1`] = `
<div>
  <form
    class="user-form"
    data-testid="user-form"
    method="post"
  >
    <h2
      class="title"
    >
      Create new user
    </h2>
    <div
      class="user-name"
    >
      <label
        class="input-label"
        for="name"
      >
        Name
      </label>
      <div
        class="input-wrapper"
      >
        <input
          class="input undefined"
          id="name"
          minlength="6"
          name="name"
          placeholder="Enter name"
          type="text"
          value=""
        />
      </div>
    </div>
    <div
      class="avatar-url"
    >
      <label
        class="input-label"
        for="avatar"
      >
        Avatar
      </label>
      <div
        class="input-wrapper"
      >
        <input
          class="input undefined"
          id="avatar"
          name="avatar"
          placeholder="Enter link avatar"
          type="url"
          value=""
        />
      </div>
    </div>
    <div
      class="button-wrapper"
    >
      <a
        class="btn-secondary cancel"
        href="/"
      >
        Cancel
      </a>
      <button
        class="btn-success"
        type="submit"
      >
        <div
          class="loader progress"
        />
      </button>
    </div>
  </form>
</div>
`;

exports[`Test UserForm component It should render correctly with initial Props 1`] = `
<div>
  <form
    class="user-form"
    data-testid="user-form"
    method="post"
  >
    <h2
      class="title"
    >
      Create new user
    </h2>
    <div
      class="user-name"
    >
      <label
        class="input-label"
        for="name"
      >
        Name
      </label>
      <div
        class="input-wrapper"
      >
        <input
          class="input undefined"
          id="name"
          minlength="6"
          name="name"
          placeholder="Enter name"
          type="text"
          value=""
        />
      </div>
    </div>
    <div
      class="avatar-url"
    >
      <label
        class="input-label"
        for="avatar"
      >
        Avatar
      </label>
      <div
        class="input-wrapper"
      >
        <input
          class="input undefined"
          id="avatar"
          name="avatar"
          placeholder="Enter link avatar"
          type="url"
          value=""
        />
      </div>
    </div>
    <div
      class="button-wrapper"
    >
      <a
        class="btn-secondary cancel"
        href="/"
      >
        Cancel
      </a>
      <button
        class="btn-success"
        type="submit"
      >
        Submit
      </button>
    </div>
  </form>
</div>
`;
